apps/gk_species_bflux.c

```diff
@@ -42,13 +134,448 @@ gk_species_bflux_rhs(gkyl_gyrokinetic_app *app, const struct gk_species *species
+      gkyl_bc_basic_advance(bflux->gfss_bc_op[b], bflux->bc_buffer, app->gk_geom->bmag);
+
+    bflux->f = gkyl_malloc(bflux->num_boundaries*sizeof(struct gkyl_array *));
+    bflux->f1 = gkyl_malloc(bflux->num_boundaries*sizeof(struct gkyl_array *));
+    bflux->fnew = gkyl_malloc(bflux->num_boundaries*sizeof(struct gkyl_array *));
+    for (int b=0; b<bflux->num_boundaries; ++b) {
+      // Allocate arrays storing moments of the boundary flux.
+      bflux->f[b] = mkarr(app->use_gpu, num_mom*app->basis.num_basis, app->local_ext.volume);
+      bflux->f1[b] = mkarr(app->use_gpu, num_mom*app->basis.num_basis, app->local_ext.volume);
+      bflux->fnew[b] = mkarr(app->use_gpu, num_mom*app->basis.num_basis, app->local_ext.volume);
+  
+      // Allocate solver.
+      int dir = bflux->boundaries_dir[b];
+      struct gkyl_range *skin_r = bflux->boundaries_edge[b]==GKYL_LOWER_EDGE? &gk_s->lower_skin[dir] : &gk_s->upper_skin[dir];
+      struct gkyl_range *ghost_r = bflux->boundaries_edge[b]==GKYL_LOWER_EDGE? &gk_s->lower_ghost[dir] : &gk_s->upper_ghost[dir];
+      bflux->flux_slvr[b] = gkyl_boundary_flux_new(dir, bflux->boundaries_edge[b], &gk_s->grid,
+        skin_r, ghost_r, gk_s->eqn_gyrokinetic, true, app->use_gpu);
+    }
+  
+    // Updater to compute the volume integral of the boundary flux moments.
+    bflux->integ_op = gkyl_array_integrate_new(&app->grid, &app->basis,
+      num_mom, GKYL_ARRAY_INTEGRATE_OP_NONE, app->use_gpu);
+    if (app->use_gpu) {
+      bflux->int_moms_local = gkyl_cu_malloc(num_mom*sizeof(double));
+      bflux->int_moms_global = gkyl_cu_malloc(num_mom*sizeof(double));
+    }
+    else {
+      bflux->int_moms_local = gkyl_malloc(num_mom*sizeof(double));
+      bflux->int_moms_global = gkyl_malloc(num_mom*sizeof(double));
+    }
+
+    for (int b=0; b<bflux->num_boundaries; ++b) {
+      bflux->intmom[b] = gkyl_dynvec_new(GKYL_DOUBLE, num_mom);
+      bflux->is_first_intmom_write_call[b] = true;
+    }
+    bflux->is_not_first_restart_write_call = true;
+
+    // Cummulative integrated moments of boundary fluxes.
+    bflux->intmom_cumm_buff = gkyl_malloc(bflux->num_boundaries*num_mom*sizeof(double));
+    for (int b=0; b<bflux->num_boundaries; ++b) {
+      for (int k=0; k<num_mom; k++)
+        bflux->intmom_cumm_buff[b*num_mom+k] = 0.0;
+    }
+  
+  }
+
+  if (!is_diagnostic) {
+    bflux->allocated_solver = true;
+
+    if (app->field->update_field && app->field->gkfield_id == GKYL_GK_FIELD_BOLTZMANN)
+      bflux->bflux_rhs_func = gk_species_bflux_rhs_solver; 
+    else
+      bflux->bflux_rhs_func = gk_species_bflux_rhs_solver_init; 
+
+    for (int d=0; d<app->cdim; ++d) {
+      for (int e=0; e<2; ++e) {
+        // Allocate solver.
+        struct gkyl_range *skin_r = e==0? &gk_s->lower_skin[d] : &gk_s->upper_skin[d];
+        struct gkyl_range *ghost_r = e==0? &gk_s->lower_ghost[d] : &gk_s->upper_ghost[d];
+        bflux->flux_slvr[2*d+e] = gkyl_boundary_flux_new(d, e, &gk_s->grid,
```

Comment:
Why not just initialize flux_slvr for all cases? How is this not an instance of code duplication? What is an example of a case where neither is initialized?